syntax = "proto3";

package protos;

import "google/protobuf/empty.proto";
import "custom_types.proto";
import "loan.proto";

// Enum для указания желаемого формата отчета
enum ReportFormat {
  DEFAULT_STRUCTURED = 0; // Стандартный структурированный ответ protobuf
  PDF = 1;
  EXCEL = 2;
  PLOTLY_JSON = 3; // JSON для графика Plotly
}

// Сообщение для возврата файла
message FileResponse {
  string filename = 1;
  bytes content = 2; // Бинарное содержимое файла
  string mime_type = 3; // Например, "application/pdf"
}

service LoanAnalyticService {
  // Рассчитывает общий портфель активных кредитов
  rpc GetTotalActiveLoanPortfolio(GetTotalActiveLoanPortfolioRequest) returns (TotalActiveLoanPortfolioResponse);

  // Анализирует распределение кредитов по статусам
  rpc GetLoanDistributionByStatus(GetLoanDistributionByStatusRequest) returns (LoanDistributionByStatusResponse);

  // Рассчитывает среднюю процентную ставку по кредитам
  rpc GetAverageLoanInterestRate(GetAverageLoanInterestRateRequest) returns (AverageLoanInterestRateResponse);

  // Прогнозирует риски невозврата кредитов (кредитный скоринг)
  rpc PredictLoanDefaultRisk(PredictLoanDefaultRiskRequest) returns (PredictLoanDefaultRiskResponse);

  // Сегментирует кредитный портфель
  rpc GetLoanPortfolioSegmentation(GetLoanPortfolioSegmentationRequest) returns (GetLoanPortfolioSegmentationResponse);
}

// --- Сообщения для запросов с выбором формата ---

message GetTotalActiveLoanPortfolioRequest {
  optional ReportFormat format = 1;
}

message GetLoanDistributionByStatusRequest {
  optional ReportFormat format = 1;
}

message GetAverageLoanInterestRateRequest {
  optional ReportFormat format = 1;
}

// --- Сообщения для ответов с поддержкой файлов ---

message TotalActiveLoanPortfolioResponse {
  message StructuredResponse {
    string total_portfolio = 1;
  }
  oneof response_data {
    StructuredResponse structured_response = 1;
    FileResponse file_response = 2;
  }
}

message LoanDistributionByStatusResponse {
  message StructuredResponse {
    map<string, int32> status_distribution = 1;
  }
  oneof response_data {
    StructuredResponse structured_response = 1;
    FileResponse file_response = 2;
  }
}

message AverageLoanInterestRateResponse {
  message StructuredResponse {
    string average_interest_rate = 1;
  }
  oneof response_data {
    StructuredResponse structured_response = 1;
    FileResponse file_response = 2;
  }
}

message PredictLoanDefaultRiskRequest {
  int32 loan_id = 1;
}

message PredictLoanDefaultRiskResponse {
  double risk_score = 1; // или float
}

enum SegmentationType {
  BY_CUSTOMER_TYPE = 0;
  BY_TERM = 1;
}

message GetLoanPortfolioSegmentationRequest {
  SegmentationType segmentation_type = 1;
  optional ReportFormat format = 2;
}

message LoanSegment {
    string segment_name = 1;
    repeated LoanModel loans = 2;
}

message GetLoanPortfolioSegmentationResponse {
    message StructuredResponse {
        repeated LoanSegment segments = 1;
    }
    oneof response_data {
        StructuredResponse structured_response = 1;
        FileResponse file_response = 2;
    }
}
